lib/portage/dep/libc.py

```diff
@@ -0,0 +1,69 @@
+# Copyright 2023 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+from portage.dep import Atom
+from portage.const import LIBC_PACKAGE_ATOM
+from portage.dbapi._expand_new_virt import expand_new_virt
+
+import portage.dbapi.porttree
+
+
+def find_libc_deps(portdb: portage.dbapi.porttree.dbapi, realized: bool = False):
+    """Finds libc package for a ROOT via portdb.
+
+    Parameters
+    ----------
+    portdb : dbapi
+        dbapi instance for portdb (for installed packages).
+    realized : bool
+        Request installed atoms rather than the installed package satisfying LIBC_PACKAGE_ATOM.
+
+    Returns
+    -------
+    list
+        List of libc packages (or atoms if realized is passed).
+    """
+
+    libc_pkgs = set()
+
+    for atom in expand_new_virt(
+        portdb,
+        LIBC_PACKAGE_ATOM,
+    ):
+        if atom.blocker:
+            continue
+
+        if not realized:
+            # Just the raw packages were requested (whatever satifies the virtual)
+            libc_pkgs.add(atom)
+            continue
+
+        # This will give us something like sys-libs/glibc:2.2, but we want to know
+        # what installed atom actually satifies that.
+        try:
+            libc_pkgs.add(portdb.match(atom)[0])
+        except IndexError:
+            continue
+
+    return libc_pkgs
+
+
+def strip_libc_deps(dep_struct, libc_deps):
+    """TODO"""
+    # We're going to just grab the libc provider for ROOT and
+    # strip out any dep for the purposes of --changed-deps.
+    # We can't go off versions, even though it'd be more precise
+    # (see below), because we'd end up with FPs and unnecessary
+    # --changed-deps results far too often.
+    #
+    # This penalizes a bit the case where someone adds a
+    # minimum (or maximum) version of libc explicitly in an ebuild
+    # without a new revision, but that's extremely rare, and doesn't
+    # feel like it changes the balance for what we prefer here.
+    for libc_dep in libc_deps:
+        for i, x in enumerate(dep_struct):
+            if isinstance(x, list):
+                strip_libc_deps(x, libc_deps)
```

Comment:
Maybe skip recursion here, since generated deps are always on the top-level.

Reply:
Oh yes, done!